Add `Joined` `Iterable`, Implements #108

dmfs opened the pull request:
This deprarecates `Flattened`. Note the new `Joined` class doesn't have a varargs constructor because this can be confusing when joining `Iterable`s of `Iterable`s.

dmfs commented:
I know that `Flatten` is the common term and it totally makes sense, but it sometimes looks a bit strange when you're operating on something specific that's not called `Iterable`.

I mean, which one makes more sense to you?
```java
new Flattened(new MultiInsertBatch(x, y), new LinkContractBatch(a,b,c));
// or
new Joined(new MultiInsertBatch(x, y), new LinkContractBatch(a,b,c));
```
I'd say the second one is more intuitive.

lemonboston commented:
Yes, I agree `Joined` can be more intuitive in that example and in everyday meaning of the words, 'join' is clearer than 'flatten'. But it is only true in code until a specific term become well-established. Like 'decorator', or even 'map', 'map' could be called 'convert' or 'transform' as well. Or `zip` for example is a quite special name as well.

And how I see it, these stream of data/event plus operations 'style' is used more and more, so it will be very common and everybody will have automatic associations for those terms.
I would expect Kotlin to spread, and if we switched many of the `iterators` library can become actually obsolete, because of the the extension functions Kotlin provides on `Sequence` and `Iterable`:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/index.html
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-iterable/index.html
Or we can use them alongside, since there are easy `asIterable()`, `asSequence()` conversions.

And I also expect reactive programming to spread further, and they use those terms as well.

lemonboston commented:
> it sometimes looks a bit strange when you're operating on something specific that's not called Iterable

I missed this point..  right, I see that `Joined` can be more intuitive in these case. But the question is still how to call it in general...     just a matter of preference. I am of course fine with `Joined`, too.

lemonboston commented:
I forgot to mention that I notice that `Flattened` `Iterator` is not marked deprecated here, only the `Iterable`, is that intentional?

dmfs commented:
@lemonboston I've rebased and separated the secondary ctor out into it's own `Expanded` class. I also added a varargs ctor for concise usage. Please re-review.
